### DISCUSSION TITLE
UIKit ListsOfState Case Study: Why the local dataSource var in the ViewController?

### DISCUSSION THREAD

**Discussion by ryanbooker** [pointfreeco/swift-composable-architecture#351] [Q&A] [answered]
I've just been looking through the UIKit case studies, and I noticed that the `ListsOfState` example have a `dataSource` var.

Why is that? Is there a particular reason? 

At first glance, it just looks like a way to call `tableView.reloadData`, and it's trivial to remove, and call `reloadData`instead, while using the `viewStore` directly in the tableView delegate calls.

**Comment by mbrandonw** [comment, marked as answer]
Hey @ryanbooker, the `dataSource` variable is what powers the `UITableViewDataSource` conformance. For example, it determines the number of rows in the table view:

https://github.com/pointfreeco/swift-composable-architecture/blob/11f215c3b6923dfcccec6d1de6a672963a0d3c6e/Examples/CaseStudies/UIKitCaseStudies/ListsOfState.swift#L56-L58

And the label for the cell at a particular row:

https://github.com/pointfreeco/swift-composable-architecture/blob/11f215c3b6923dfcccec6d1de6a672963a0d3c6e/Examples/CaseStudies/UIKitCaseStudies/ListsOfState.swift#L65

To me it seems necessary to implement the data source protocol. Could you explain a bit more about how it's trivial to remove? I may be missing something.

**Reply by ryanbooker** [reply]
I think my question boils down to, why not:
```Swift
self.viewStore.publisher.counters
  .sink(receiveValue: { [weak self] _ in self?.tableView.reloadData() })
  .store(in: &cancellables)
```
and in the `UITableViewDataSource`
```Swift
override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int { 
  viewStore.counters.count 
} 
```
and
```Swift
cell.textLabel?.text = "\(viewStore.counters[indexPath.row].count)"
```
I could be missing something too… I'm still getting my head around all this. :)

**Reply by mbrandonw** [reply]
Oh, haha. Yes, I suppose it could be! I'm not sure why we did it like that. I guess we just felt it was more recognizable to iOS devs to see an explicit data source ¯\\_(ツ)_/¯

**Reply by stephencelis** [reply]
I don't think you're missing anything 😅. I think it was just a pre-TCA habit of having a "data source" around. Looks like we can simplify things: https://github.com/pointfreeco/swift-composable-architecture/pull/365

**Comment by abdelmajidrajad** [comment]
I also want to ask @Brandon Williams <mbw234@gmail.com>  about the best
approach to take to have a viewStore for each cell on UIKit.

having a setter like: cell.viewStore = viewStore is an ultimate solution ?

Thanks for taking time for this.

**Reply by mbrandonw** [reply]
Yeah that seems fine to me. It might also be possible to cook up a `Store.forEach` operator that mimics the `ForEachStore` view in SwiftUI (we already have `Store.ifLet` that mimics `IfLetStore`), which would allow you to pass a `Store` to each cell and then turn it into a `ViewStore` inside the cell. But we haven't given much thought to that yet.

**Reply by mluisbrown** [reply]
Hi @mbrandonw, two of my co-workers (@gtsiap and @ekoutkias) created a version of `Store.forEach` for use in our project at XM where we are using the ReactiveSwift TCA with UIKit. It uses ReactiveSwift but should be easy to convert to Combine

```swift
import ComposableArchitecture
import Foundation
import ReactiveSwift

extension Store where State: Equatable {
    public func forEach<LocalState: Equatable, LocalAction>(
        state: @escaping (State) -> [LocalState],
        action: @escaping ((Int, LocalAction)) -> Action
    ) -> [Store<LocalState, LocalAction>] {
        let scopedStore = scope(state: state)
        let viewStore = ViewStore(scopedStore)
        let localStores = zip(viewStore.indices, viewStore.state).map { index, element in
            scopedStore.scope(
                state: { index < $0.endIndex ? $0[index] : element },
                action: { action((index, $0)) }
            )
        }
        return localStores
    }

    public func forEach<LocalState: Equatable, LocalAction>(
        state: @escaping (State) -> [LocalState],
        action: @escaping (LocalAction) -> Action
    ) -> [Store<LocalState, LocalAction>] {
        let scopedStore = scope(state: state)
        let viewStore = ViewStore(scopedStore)
        let localStores = zip(viewStore.indices, viewStore.state).map { index, element in
            scopedStore.scope(
                state: { index < $0.endIndex ? $0[index] : element },
                action: { action($0) }
            )
        }
        return localStores
    }
}
```

**Comment by abdelmajidrajad** [comment]
Thanks Brandon.
